src/Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv/Internal/LibuvConnection.cs

```diff
@@ -175,27 +175,34 @@ private static void ReadCallback(UvStreamHandle handle, int status, object state
             _currentWritableBuffer = null;
             _bufferHandle.Free();
 
-            if (flushTask?.IsCompleted == false)
-            {
-                Log.ConnectionPause(ConnectionId);
-                StopReading();
-
-                var result = await flushTask.Value;
-                // If the reader isn't complete then resume
-                if (!result.IsCompleted && !result.IsCancelled)
-                {
-                    Log.ConnectionResume(ConnectionId);
-                    StartReading();
-                }
-            }
-
             if (!normalRead)
             {
                 _connectionContext.Abort(error);
 
                 // Complete after aborting the connection
                 Input.Complete(error);
             }
+            else if (flushTask?.IsCompleted == false)
+            {
+                // We wrote too many bytes too the reader so pause reading and resume when
+                // we hit the low water mark
+                _ = ApplyBackpressureAsync(flushTask.Value);
+            }
+        }
+
+        private async Task ApplyBackpressureAsync(WritableBufferAwaitable flushTask)
```

Comment:
nit: since this also releases backpressure maybe name this RegulateBackpressureAsync